aten/src/ATen/native/sparse/cuda/cutlass/README.md

```diff
@@ -0,0 +1,29 @@
+This directory contains files from CUTLASS 3.1 source tree, modified
```

Comment:
Is it possible to keep these modified CUTLASS sources temporarily and submit the modifications upstream, then remove them from the pytorch tree when a release is minted included the additional functionality? 

I worry about having to indefinitely support them breaking in subtle ways with updates to our pinned version of CUTLASS in the future. 

It would be nice to have a plan for their eventual removal.

Reply:
At the moment, CUTLASS supports adding bias vector to rows of a row-major output matrix, and to columns of a column-major output matrix.  As we assume that weight matrix is sparse, and as the sparse matrix has to be first operand of CTULASS sparse GEMM, this new linear operator is calculated as $(Wx^{T}+b^{T})^T$ (instead of $xW^{T}+b$, as usual); so it turns out we need to add bias vector to columns of output matrix.  In principle, we can choose any of these two layouts for the output matrix, but at the moment CUTLASS supports only row-major output for sparse GEMM.   Thus this hack.  I've opened an issue at CUTLASS GitHub, and was told creating customized PredicatedTileIterator class, outside of the CUTLASS, is the way; the rest of files are there only as, even CUTLASS is a highly templated library, it's actually cumbersome to combine at will.  So - I worry about the same thing (moreover, the fact that CUTLASS is not processing files through any kind of linter, and that we do, makes remembering what was actually changed from the original files somewhat problematic), but for now the plan will have to be to keep an eye both on updates of our pinned version, and the CUTLASS repo, to make changes here is updated of our pinned version break something, and on the other side to hope CUTLASS developers will provide some means to simplify this.